contracts/Minter.sol

```diff
@@ -0,0 +1,151 @@
+  /// @notice Maximum number of tokens minted, 3B (with 18 decimal places)
+  uint256 public cap = 3000000000000000000000000000;
+
+  /// @notice Struct to store information about each pool
+  struct Pool {
+    address receiver;
+    uint256 amountPerBlock;
+  }
+  /// @notice Array to store all pools
+  Pool[] public pools;
+
+  /// @notice Emitted when mint() is called
+  event Mint(address receiver, uint256 amount);
+  /// @notice Emitted when pool is added
+  event PoolAdded(address newReceiver, uint256 newAmount);
+  /// @notice Emitted when pool is updated
+  event PoolUpdated(uint256 index, address newReceiver, uint256 newAmount);
+  /// @notice Emitted when pool is removed
+  event PoolRemoved(uint256 index, address receiver, uint256 amount);
+  /// @notice Emitted when admin address is updated
+  event UpdateAdmin(address admin, address newAdmin);
+
+  /// @notice Modifier to allow only admin to call certain functions
+  modifier onlyAdmin(){
+    require(msg.sender == admin, '!admin');
+    _;
+  }
+
+  /**
+   * @notice Constructor of new minter contract
+   * @param newToken Address of the token to mint
+   * @param startBlock Initial lastMint block
+   * @param newAdmin Initial admin address
+   */
+  constructor(address newToken, uint256 startBlock, address newAdmin){
+    require(startBlock >= block.number, "Start block must be above current block");
+    require(newToken != address(0), 'Token cannot be address 0');
+    require(newAdmin != address(0), 'Admin cannot be address 0');
+
+    token = newToken;
+    lastMintBlock = startBlock;
+    admin = newAdmin;
+
+    emit UpdateAdmin(address(0), newAdmin);
+  }
+
+  /**
+   * @notice Mint tokens to all pools, can be called by anyone
+   */
+  function mint() public {
+    uint256 mintDifference = block.number - lastMintBlock;
+
+    for (uint256 i = 0; i < pools.length; i++){
+      uint256 amount = pools[i].amountPerBlock.mul(mintDifference);
+
+      if(totalMinted + amount >= cap && totalMinted != cap){
+        amount = cap.sub(totalMinted);
+      }
+      require(totalMinted.add(amount) <= cap, "Cap reached");
+
+      totalMinted = totalMinted.add(amount);
+      IERC20(token).mint(pools[i].receiver, amount);
+
+      emit Mint(pools[i].receiver, amount);
+    }
+
+    lastMintBlock = block.number;
+  }
+
+  /**
+   * @notice Add new pool, can be called by admin
+   * @param newReceiver Address of the receiver
+   * @param newAmount Amount of tokens per block
+   */
+  function addPool(address newReceiver, uint256 newAmount) external onlyAdmin {
+    pools.push(Pool(newReceiver, newAmount));
+    emit PoolAdded(newReceiver, newAmount);
+  }
+
+  /**
+   * @notice Update pool, can be called by admin
+   * @param index Index in the array of the pool
+   * @param newReceiver Address of the receiver
+   * @param newAmount Amount of tokens per block
+   */
+  function updatePool(uint256 index, address newReceiver, uint256 newAmount) external onlyAdmin {
```

Comment:
Given the sensitivity of what's going on here (minting a token with real economic value out to accounts not controlled by Splinterlands) it's advisable to perform a sanity check for `newAmount`, both here and in the `addPool` method, i.e. make sure the amount falls within a configured min/max range. This can be a constant or configurable by the admin. In particular, you want to avoid an error by the admin with decimals that could result in the minting of millions or billions of tokens unexpectedly.

Reply:
Fixed, maximum amount per block can be updated by admin: https://github.com/steem-monsters/sps-minter/commit/644429cec057b00ea4a3d7f212e8cf2b6b931129 & https://github.com/steem-monsters/sps-minter/commit/4d6646c6a13848bfd1139270a948c5ba54a3f5e5 & https://github.com/steem-monsters/sps-minter/commit/fe6db2a5b0d621c7bc3690352165a76eca89bb6c

Reply:
I want to confirm the intention here is to cap the amount that can be minted per block *to a single pool*, that is correct, right? That is what my comment suggested and it is what the code seems to do, but the naming of the variable is confusing. At first read, it seems like this parameter would be the *total* SPS that could possibly be minted per block. I recommend changing the name of the parameter to `maxToPoolPerBlock` or something similar. I also recommend updating the comment above the parameter to make this intention clear.

Separately, this brings up another question. Should there be a *total* mintable SPS cap per block? This could be configurable or hardcoded. It's probably more of a "product" question for @MattyIce than it is a technical question.

Reply:
well there's a limit per pool, and there's a limit on the number of pools, so that means that there is a cap on the total mintable SPS per block, right?

Reply:
maxPerBlock renamed to maxToPoolPerBlock: https://github.com/steem-monsters/sps-minter/commit/1ec5b1717295dbf795456ce7b5fde3d4ab9db8f4

I also think we don't need a separate "total mintable per block" cap since both the number of pools and the amount per pool are already limited.

Reply:
Sounds good, looks good 👌